src/main/java/domain/Janggi.java

```diff
@@ -0,0 +1,164 @@
+            units.remove(destination);
+        }
+        units.remove(pick);
+        units.put(destination, pickedUnit);
+        switchTurn();
+    }
+
+    private boolean canMove(Position pick, Position destination) {
+        List<Route> movableRoutes = findMovableRoutesFrom(pick);
+        return movableRoutes.stream()
+                .map(route -> route.searchDestination(pick))
+                .anyMatch(position -> position.equals(destination));
+    }
+
+    private void switchTurn() {
+        turn = turn.getOpposite();
+    }
+
+    public List<Route> findMovableRoutesFrom(Position pick) {
+        if (isEmptyPosition(pick)) {
+            throw new IllegalArgumentException(EMPTY_POINT_EXCEPTION);
+        }
+        Unit pickedUnit = units.get(pick);
+        if (pickedUnit.getTeam() != turn) {
+            throw new IllegalArgumentException(PICK_OPPOSITE_UNIT_EXCEPTION);
+        }
+
+        List<Route> totalRoutes = pickedUnit.calculateRoutes(pick);
+        totalRoutes = filterRoutesByUnitType(pickedUnit, pick, totalRoutes);
+        if (pickedUnit.getType() == UnitType.CANNON) {
+            return totalRoutes;
+        }
+        return filterBlockedRoutes(pick, totalRoutes);
+    }
+
+    private List<Route> filterRoutesByUnitType(Unit pickedUnit, Position pick, List<Route> totalRoutes) {
+        UnitType type = pickedUnit.getType();
+        if (type == UnitType.CANNON) {
+            return totalRoutes.stream()
+                    .filter(route -> canCannonJump(pick, route))
+                    .toList();
+        }
+        if (type == UnitType.SOLDIER) {
+            return filterSoldierMoves(pick, pickedUnit, totalRoutes);
+        }
+        return totalRoutes;
+    }
+
+    private List<Route> filterSoldierMoves(Position pick, Unit pickedUnit, List<Route> totalRoutes) {
+        if (pickedUnit.getTeam() == Team.HAN) {
+            return totalRoutes.stream()
+                    .filter(route -> route.getPositions().getFirst().getY() >= pick.getY())
+                    .toList();
+        }
+        return totalRoutes.stream()
+                .filter(route -> route.getPositions().getFirst().getY() <= pick.getY())
+                .toList();
+    }
+
+    private boolean canCannonJump(Position current, Route route) {
+        Position endPoint = route.searchDestination(current);
+        if (!isEmptyPosition(endPoint)) {
+            Unit endUnit = units.get(endPoint);
+            if (endUnit.getType() == UnitType.CANNON) {
+                return false;
+            }
+        }
+
+        int count = 0;
+        for (Position position : route.getPositionsExceptDestination(current)) {
+            if (isEmptyPosition(position)) {
+                continue;
+            }
+            Unit unit = units.get(position);
+            if (unit.getType() == UnitType.CANNON) {
+                return false;
+            }
+            count++;
+        }
+        return (count == 1);
+    }
```

Comment:
조금 중복이될 수도 있지만, 지금 코드는 한번에 여러 계산을하려다보니 조금 복잡하게 느껴지는 것 같아요. 

아래와 같이 변경해보는건 어떻게 생각하시나요? 😃 (혹시 해당 피드백을 반영하신다면 의도를 드러낼 수 있는 더 좋은 네이밍을 고려해주시면 좋을 것 같습니다 😉)

```suggestion
            var positions = route.getPositionsExceptDestination(current);
        var hasCannon = positions.stream()
            .filter(units::containsKey)
            .map(units::get)
            .anyMatch(unit -> unit.getType() == UnitType.CANNON);

        return !hasCannon && onlyOne(positions);
    }

    private boolean onlyOne(List<Position> positions) {
        var count = positions.stream()
            .filter(units::containsKey)
            .map(units::get)
            .count();
        return count == 1;
    }
```